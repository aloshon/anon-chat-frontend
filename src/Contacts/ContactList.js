import React, {useEffect, useContext} from "react";
import { useDispatch, useSelector, shallowEqual } from "react-redux";
import Contact from "./Contact";
import AddContactForm from "../Forms/AddContactForm";
import Container from "react-bootstrap/Container";
import {getContactList} from "../Actions/actionCreators";
import UserContext from "../UserContext";

/**
 * ContactList component displays list of contacts
 * User needs to be logged in
 */
const ContactList = () => {
    const contacts = useSelector(state => state.contacts, shallowEqual);
    const dispatch = useDispatch();
    const {user} = useContext(UserContext);

    // Get the user's contactList on render after user is defined
    useEffect(() =>{
      if(user && contacts.length === 0){
          getContactList(dispatch, user.contactList);
      }
  }, [user, dispatch, contacts]);

  // If user is not logged in or loaded yet, return loading
  while(!user) return <h1>Loading...</h1>;

    // marginBottom style is to keep the last contact
    // from touching the very bottom of the screen
    return (
        <>
      <h1 className="my-3">Contact List</h1>
      <AddContactForm />
      {contacts.length !== 0
        ? (
          <Container style={{ marginBottom: '60px' }}>
              {contacts.map(c =>
                <Contact 
                key={c.user_id}
                user_id={c.user_id}
                nickname={c.nickname}
                username={c.username}
                />
              )}
          </Container>
        )
        : <h6>No contacts yet...</h6>
      }
    </>
    );
}

export default ContactList;