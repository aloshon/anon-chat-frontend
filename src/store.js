import  { composeWithDevTools } from "redux-devtools-extension";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import autoMergeLevel2 from "redux-persist/lib/stateReconciler/autoMergeLevel2";

import thunk from "redux-thunk";
import rootReducer from "./Reducers/rootReducer";
import { createStore, applyMiddleware } from "redux";

// only keep the darkMode reducer persistent in local storage
const persistConfig = {
  key: "root",
  whitelist:['darkMode'],
  storage,
  stateReconciler: autoMergeLevel2
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(
  persistedReducer,
  composeWithDevTools(
    applyMiddleware(thunk),
  )
);

export const persistedStore = persistStore(store);