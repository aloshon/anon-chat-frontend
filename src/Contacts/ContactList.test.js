import { render, screen } from '@testing-library/react';
import ContactList from './ContactList';
import { Provider } from "react-redux";
import { store } from "../store"
import {BrowserRouter} from 'react-router-dom';
import UserContext from "../UserContext";

describe("Renders list of contacts", () => {
    const user = {
        id: 1,
        username: "user1",
        isAdmin: false,
        blockList: []
    }

    const {asFragment} = render(
      <Provider store={store}>
        <BrowserRouter>
            <UserContext.Provider value={{user}}>
                <ContactList />
            </UserContext.Provider>
        </BrowserRouter>
      </Provider>
    );
    test("Contact List title is rendered", () => {
      const title = screen.getByText("Contact List");
      expect(title).toBeInTheDocument();
    });
  
    test("matches snapshot", () => {
      expect(asFragment()).toMatchSnapshot();
    });
    
  });