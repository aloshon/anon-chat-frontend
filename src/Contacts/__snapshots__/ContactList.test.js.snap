// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders list of contacts matches snapshot 1`] = `<DocumentFragment />`;
